presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java

```diff
@@ -0,0 +1,235 @@
+        public PlanNodeCostEstimate visitJoin(JoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node,
+                    node.getLeft(),
+                    node.getRight(),
+                    node.getDistributionType().orElse(JoinNode.DistributionType.PARTITIONED).equals(JoinNode.DistributionType.REPLICATED));
+        }
+
+        private PlanNodeCostEstimate calculateJoinCost(PlanNode join, PlanNode probe, PlanNode build, boolean replicated)
+        {
+            int numberOfNodesMultiplier = replicated ? numberOfNodes : 1;
+
+            PlanNodeStatsEstimate probeStats = getStats(probe);
+            PlanNodeStatsEstimate buildStats = getStats(build);
+            PlanNodeStatsEstimate outputStats = getStats(join);
+
+            double cpuCost = probeStats.getOutputSizeInBytes() +
+                    buildStats.getOutputSizeInBytes() * numberOfNodesMultiplier +
+                    outputStats.getOutputSizeInBytes();
+
+            if (replicated) {
+                // add the cost of a local repartitioning of build side copies
+                // cost of the repartitioning of a single data copy has been already added in calculateExchangeCost
+                cpuCost += buildStats.getOutputSizeInBytes() * (numberOfNodesMultiplier - 1);
+            }
+
+            double memoryCost = buildStats.getOutputSizeInBytes() * numberOfNodesMultiplier;
+
+            return new PlanNodeCostEstimate(cpuCost, memoryCost, 0);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitExchange(ExchangeNode node, Void context)
+        {
+            return calculateExchangeCost(numberOfNodes, getStats(node), node.getType(), node.getScope());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitTableScan(TableScanNode node, Void context)
+        {
+            // TODO: add network cost, based on input size in bytes? Or let connector provide this cost?
+            return cpuCost(getStats(node).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitValues(ValuesNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitSemiJoin(SemiJoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node,
+                    node.getSource(),
+                    node.getFilteringSource(),
+                    node.getDistributionType().orElse(SemiJoinNode.DistributionType.PARTITIONED).equals(SemiJoinNode.DistributionType.REPLICATED));
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitLimit(LimitNode node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputSizeInBytes());
```

Comment:
It does not seem right. IMO `ZERO` is closer to truth. 

`LimitNode` seems to be tricky, it has a cost feedback loop to cost of its sources nodes. Notice even if you have a large cost of scan, having limit on top of it will make the cost of scan lower as not all the data has to be read.

Reply:
as discussed, I didn't change anything, but added explanatory comment to the code